control/controlclient/hostinfo_windows.go

```diff
@@ -5,26 +5,16 @@
 package controlclient
 
 import (
-	"os/exec"
-	"strings"
-	"syscall"
+	"fmt"
+
+	"golang.org/x/sys/windows"
 )
 
 func init() {
 	osVersion = osVersionWindows
 }
 
 func osVersionWindows() string {
-	cmd := exec.Command("cmd", "/c", "ver")
-	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
-	out, _ := cmd.Output() // "\nMicrosoft Windows [Version 10.0.19041.388]\n\n"
-	s := strings.TrimSpace(string(out))
-	s = strings.TrimPrefix(s, "Microsoft Windows [")
-	s = strings.TrimSuffix(s, "]")
-
-	// "Version 10.x.y.z", with "Version" localized. Keep only stuff after the space.
-	if sp := strings.Index(s, " "); sp != -1 {
-		s = s[sp+1:]
-	}
-	return s // "10.0.19041.388", ideally
+	major, minor, build := windows.RtlGetNtVersionNumbers()
+	return fmt.Sprintf("%d.%d.%d", major, minor, build)
```

Comment:
this only has 3 numbers. The code before had 4 numbers.

https://www.geoffchappell.com/studies/windows/win32/ntdll/api/ldrinit/getntversionnumbers.htm says "The optional BuildNumber argument gives the address of a variable that is to receive a number that describes the build. [...] The low 16 bits are the build number as commonly understood. The high four bits of the number distinguish free and checked builds."

Not sure whether that helps us map back to the same output as before.

Reply:
As far as I know, the first three numbers (major, minor and build), uniquely identify the version. There are no two Windows versions where only the fourth number (revision) differs. ([Example](https://www.lifewire.com/windows-version-numbers-2625171))
Also required `windows.RtlGetNtVersionNumbers` API supports only major.minor.build. Mentioned low/high bits thing doesn't contain revision info - it's only [free/checked build](https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/checked-and-free-build-differences) info.
If all four numbers are actually needed, then the simplest way is to keep the existing `ver` output parsing method, maybe only simplify the parsing (use regex instead of indexof)?

Reply:
Actually, I want all four numbers back.

I spent the evening debugging an issue for a user running an old version of Windows (https://github.com/tailscale/tailscale/issues/716#issuecomment-799931271) and then going to to https://en.wikipedia.org/wiki/Windows_10_version_history_(version_1809) and reading all the revision notes in the footnotes of that wikipedia page.

I'm going to revert this for now for Tailscale 1.6. We can land it again later when we figure out how to get the fourth numbers.